autoload/crystalline.vim

```diff
@@ -461,6 +525,11 @@ function! crystalline#apply_current_theme() abort
   catch /^Vim\%((\a\+)\)\=:E118/
     " theme does not use autoload function
   endtry
+
+  call crystalline#generate_theme(g:crystalline_sep_hi_groups)
+  " Empty out after adjusting separator HGs
+  let g:crystalline_sep_hi_groups = {}
```

Comment:
This wasn't quite what I was thinking. However, I've done some testing and this works, but I don't know why, both on colorscheme change and on calling `set_theme`. It even works if I comment out this whole block of code.

What I was thinking was that `highlight clear (generated sep group 1) (generated sep group 2) (etc...)` would be called here so that when you check for `hlexists` it would re-generate the highlight groups.

It makes sense that changing the colorscheme would work, since colorschemes normally clear out all highlight groups before setting them. However, I don't know why this works on `set_theme`. Let me know if you understand why this is working since it seems like I accidentally made you add unnecessary code. I will also keep doing my own digging.

Reply:
If you want to test dynamic changing of the colorscheme, you can do it with:

```
call crystalline#set_theme("gruvbox")
set background=dark
set background=light
```

This should dynamically switch between light/dark.

Regular calls to `set_theme` can be tested just by calling it with different colorschemes.

Reply:
> However, I don't know why this works on `set_theme`. Let me know if you understand why this is working since it seems like I accidentally made you add unnecessary code.

What is the 'this' in 'this works'? Changing separators when the colorscheme changes?

Reply:
>What is the 'this' in 'this works'? Updating separators when the colorscheme changes?

"This" as in updating separators when `set_theme` is called.

Reply:
It appears that my approach of clearing out highlight groups won't work - `hlexists` will return true if a highlight group was ever defined. However `generate_theme` also won't work since `g:crystalline_hi_sep_groups` will contain the old attributes here.

Reply:
I tried also just re-creating themes using new values, but that also has its own problems - say theme A decides to not generate `CrystallineFooToBar` and theme B does. Say theme A is our initial theme. When the statusline is generated, `CrystallineFooToBar` will be created and registered as a generated highlight group. When the theme is changed to theme B, we will overwrite `CrystallineFooToBar`.

What we really want is a way to __delete__ highlight groups so `hlexists` will return false but I don't think it exists. Might have to fall back on `getcompletion` rather than `hlexists` or change something about how we require themes to be set.

Reply:
Yeah, I just came to the same conclusion. I mean, there's always this monstrosity:

```vim
function HLExists(hi) abort
  if !hlexists(a:hi)
    return 0
  endif
  redir => l:hlstatus
  highlight a:hi
  redir END
  return l:hlstatus =~ 'cleared'
endfunction
```

But… ugh.

Reply:
> say theme A decides to not generate CrystallineFooToBar and theme B does.

Wait a second—if this PR actually ends up merged, why would a theme specify highlight groups for separators itself? Are you talking about name collisions between HGs statically defined by the theme and dynamically generated HGs?

Reply:
You're right actually, a theme wouldn't want to manually specify highlight groups. It also shouldn't be a problem for custom non-separator highlight groups because it should be up to the theme to clear those out.

Reply:
In that case this approach should work:

- Instead of checking if highlight groups exist using `hlexists`, check if they are set in the `g:crystalline_sep_hi_groups` object
- Clear out `g:crystalline_sep_hi_groups` when `apply_current_theme` is called (you're already doing this, just saying we shouldn't have to call `highlight clear` or anything like I thought)

Reply:
That did the trick!